x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java

```diff
@@ -0,0 +1,526 @@
+                // encoded string, need to re-encode
+                valueAsString = ((BytesRef) value).utf8ToString();
+            } else {
+                throw new IllegalArgumentException("Illegal value type: " + value.getClass() + ", value: " + value);
+            }
+            return encodeVersion(valueAsString).bytesRef;
+        }
+
+        @Override
+        public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName) {
+            failIfNoDocValues();
+            return new SortedSetOrdinalsIndexFieldData.Builder(name(), VersionScriptDocValues::new, CoreValuesSourceType.BYTES);
+        }
+
+        @Override
+        public Object valueForDisplay(Object value) {
+            if (value == null) {
+                return null;
+            }
+            return VERSION_DOCVALUE.format((BytesRef) value);
+        }
+
+        @Override
+        public DocValueFormat docValueFormat(@Nullable String format, ZoneId timeZone) {
+            if (format != null) {
+                throw new IllegalArgumentException("Field [" + name() + "] of type [" + typeName() + "] does not support custom formats");
+            }
+            if (timeZone != null) {
+                throw new IllegalArgumentException(
+                    "Field [" + name() + "] of type [" + typeName() + "] does not support custom time zones"
+                );
+            }
+            return VERSION_DOCVALUE;
+        }
+
+        @Override
+        public Query rangeQuery(Object lowerTerm, Object upperTerm, boolean includeLower, boolean includeUpper, QueryShardContext context) {
+            if (context.allowExpensiveQueries() == false) {
+                throw new ElasticsearchException(
+                    "[range] queries on [version] fields cannot be executed when '"
+                        + ALLOW_EXPENSIVE_QUERIES.getKey()
+                        + "' is set to false."
+                );
+            }
+            failIfNotIndexed();
+            BytesRef lower = lowerTerm == null ? null : indexedValueForSearch(lowerTerm);
+            BytesRef upper = upperTerm == null ? null : indexedValueForSearch(upperTerm);
+            byte[] lowerBytes = lower == null ? MIN_VALUE : Arrays.copyOfRange(lower.bytes, lower.offset, 16);
+            byte[] upperBytes = upper == null ? MAX_VALUE : Arrays.copyOfRange(upper.bytes, upper.offset, 16);
+
+            // point query on the 16byte prefix
+            Query pointPrefixQuery = BinaryPoint.newRangeQuery(name(), lowerBytes, upperBytes);
+
+            Query termQuery = new TermRangeQuery(
+                name(),
+                lowerTerm == null ? null : lower,
+                upperTerm == null ? null : upper,
+                includeLower,
+                includeUpper
+            );
+
+            return new BooleanQuery.Builder().add(new BooleanClause(pointPrefixQuery, Occur.MUST))
+                .add(new BooleanClause(termQuery, Occur.MUST))
+                .build();
```

Comment:
The two queries that you mixed are totally independent so the resulting boolean query is just more costly than a plain `TermRangeQuery`. 
The idea behind using a prefix query on `Points` is that you can use the doc values to validate the matches. You can check how the `wildcard` field [does](https://github.com/elastic/elasticsearch/blob/2c501a580c857c232d96195b410040f063b0169a/x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java#L726)  since it's very similar. The `AutomatonQueryOnBinaryDv` leverages the `TwoPhaseIterator` to only validate documents found by other filter (the point range query approximation). 
As I said in previous comments, I think we need to make a choice. Relying on `Points` or `Postings` is ok but not both.

Reply:
Thanks, I looked into this and hope I understood things right. I pushed cfe2dba which follows a simiar approach as AutomatonQueryOnBinaryDv but works on SortedSet doc values.

Reply:
After talking again about this I changed the range validation query again in eaa1ef8. Since we also talked about not using Points at all I will use this version when comparing to a solution that doesn't use Points at all (a simple TermRangeQuery)